lib/cn/csched_sp3_work.c

```diff
@@ -223,13 +223,31 @@ sp3_work_wtype_idle(
     *action = CN_ACTION_COMPACT_KV;
 
     kvsets = cn_ns_kvsets(&tn->tn_ns);
+    if (kvsets < 2)
+        return 0;
 
     /* Keep idle index nodes fully compacted to improve scanning
      * (e.g., mongod index nodes that rarely change after load).
      */
     if (cn_ns_vblks(&tn->tn_ns) < kvsets) {
+        const ulong keys_max = (ulong)thresh->lcomp_split_keys << 21;
+
+        /* Skip oldest kvsets with enormous key counts.
+         */
+        for (le = *mark; le; le = list_prev_entry_or_null(le, le_link, head)) {
+            const struct kvset_stats *stats = kvset_statsp(le->le_kvset);
+
+            if (stats->kst_keys < keys_max)
+                break;
+
+            kvsets--;
+        }
+
+        kvsets = (kvsets > 1) ? kvsets : 0;
         *rule = CN_RULE_IDLE_INDEX;
-        return kvsets;
+        *mark = le;
+
+        return min_t(uint, kvsets, thresh->lcomp_runlen_max);
```

Comment:
This should in turn probably be a uint64_t?

Reply:
No, all these work funcs return a uint, the number of kvsets in a node will never be anywhere larger than say 100, and that's if things have gone horribly wrong.

Reply:
Then why is number of kvsets a 32 bit int?

Reply:
You mean 64-bit?  Probably just not well-considered...

Reply:
Oh, I see, for some reason cn_ns_kvsets() returns u64, but kst_kvsets is u32, dunno, probably a copy-n-paste job from long ago?